File: backend/app/schemas/trips.py
from __future__ import annotations

import datetime as dt

from pydantic import BaseModel, Field


class PlaceOut(BaseModel):
    id: str
    order_index: int
    name: str
    description: str | None = None
    category: str | None = None
    lat: float
    lon: float
    address: str | None = None
    duration_minutes: int | None = None


class DayOut(BaseModel):
    id: str
    day_number: int
    date: dt.date | None = None
    title: str | None = None
    places: list[PlaceOut] = Field(default_factory=list)


class TripBase(BaseModel):
    title: str | None = None
    destination: str = Field(min_length=1, max_length=200)
    start_date: dt.date | None = None
    end_date: dt.date | None = None
    travelers: int = 1
    interests: list[str] = Field(default_factory=list)


class TripCreate(TripBase):
    pass


class TripPatch(BaseModel):
    title: str | None = None
    is_public: bool | None = None


class TripSummary(BaseModel):
    id: str
    title: str | None
    destination: str
    start_date: dt.date | None
    end_date: dt.date | None
    created_at: dt.datetime
    is_public: bool


class TripOut(TripBase):
    id: str
    summary: str | None = None
    is_public: bool
    days: list[DayOut] = Field(default_factory=list)
    created_at: dt.datetime
    updated_at: dt.datetime

File: frontend/src/api/trips.ts
import { api } from "./client";

export type Place = {
  id: string;
  order_index: number;
  name: string;
  description: string | null;
  category: string | null;
  lat: number;
  lon: number;
  address: string | null;
  duration_minutes: number | null;
};

export type Day = {
  id: string;
  day_number: number;
  date: string | null;
  title: string | null;
  places: Place[];
};

export type Trip = {
  id: string;
  title: string | null;
  destination: string;
  start_date: string | null;
  end_date: string | null;
  travelers: number;
  interests: string[];
  summary: string | null;
  is_public: boolean;
  days: Day[];
  created_at: string;
  updated_at: string;
};

export type TripSummary = {
  id: string;
  title: string | null;
  destination: string;
  start_date: string | null;
  end_date: string | null;
  created_at: string;
  is_public: boolean;
};

export const createTrip = (body: {
  destination: string;
  start_date?: string | null;
  end_date?: string | null;
  travelers?: number;
  interests?: string[];
  title?: string;
}) => api.post<Trip>("/trips", body).then((r) => r.data);

export const listTrips = () =>
  api.get<TripSummary[]>("/trips").then((r) => r.data);

export const getTrip = (id: string) =>
  api.get<Trip>(`/trips/${id}`).then((r) => r.data);

export const patchTrip = (
  id: string,
  body: Partial<{ title: string; is_public: boolean }>,
) => api.patch<Trip>(`/trips/${id}`, body).then((r) => r.data);

export const deleteTrip = (id: string) =>
  api.delete(`/trips/${id}`).then((r) => r.data);

export const publicTrip = (id: string) =>
  api.get<Trip>(`/public/trips/${id}`).then((r) => r.data);
